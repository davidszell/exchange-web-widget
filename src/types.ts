export type ExchangeRatesType = {
    base: string,
    rates: {
        [id: string]: number
    }
}

export type WalletType = {
    name: string,
    balance: number,
    longName: string,
    symbol: string
}