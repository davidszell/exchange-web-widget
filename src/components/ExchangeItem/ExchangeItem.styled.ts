import styled from 'styled-components';
import { Button } from '../Exchange/Exchange.styled';

export const Container = styled.div`
    padding: 10px;
    border: 1px solid #ccc;
    border-radius: 5px;
    display: flex;
    flex-direction: column;
    gap: 10px;
`;

export const CurrencyContainer = styled.div`
    display: flex;
    flex-direction: row;
    gap: 10px;
`;

export const CurrencyButton = styled(Button)`
    &::after {
        content: '‚ÆÅ';
        margin-left: 10px;
    }
`;

export const Amount = styled.input`
    flex: 1;
`;

export const Balance = styled.p`
    font-size: 0.8em;
    margin: 0;
    color: #777;
`;
