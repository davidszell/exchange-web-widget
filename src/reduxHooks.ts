import { TypedUseSelectorHook, useDispatch, useSelector } from "react-redux";
import { AnyAction } from "redux";
import { ThunkDispatch } from "redux-thunk";
import { RootState } from "./store";

export const useAppDispatch = (): ThunkDispatch<RootState, any, AnyAction> => useDispatch<ThunkDispatch<RootState, any, AnyAction>>()
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector