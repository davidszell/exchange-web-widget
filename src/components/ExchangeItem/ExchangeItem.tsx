import React, { useState } from "react";
import { WalletType } from "../../types";
import Modal from "../Modal";
import WalletPicker from "../WalletPicker";
import { Amount, Balance, Container, CurrencyButton, CurrencyContainer } from "./ExchangeItem.styled";

type ExchangeItemProps = {
    wallet: WalletType,
    handleWalletChange: Function,
    amount: number,
    handleAmountChange: Function

}

const ExchangeItem = ({ wallet, handleWalletChange, amount, handleAmountChange }: ExchangeItemProps): JSX.Element => {
    const [showWalletPicker, setShowWalletPicker] = useState<boolean>(false);

    const handleOnChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        handleAmountChange(event.target.value);
    }

    const handleShowWalletPicker = () => {
        setShowWalletPicker(true);
    }

    const handleHideWalletPicker = () => {
        setShowWalletPicker(false);
    }

    const handleWalletPicked = (walletName: string) => {
        handleWalletChange(walletName);
        setShowWalletPicker(false);
    }

    return (
        <Container>
            <CurrencyContainer>
                <CurrencyButton onClick={handleShowWalletPicker}>{wallet.name}</CurrencyButton>
                <Amount type="number" placeholder="0" value={amount} onChange={handleOnChange} ></Amount>
            </CurrencyContainer>
            <Balance>Balance: {wallet.symbol}{wallet.balance}</Balance>
            {showWalletPicker ? (
                <Modal handleClose={handleHideWalletPicker}>
                    <WalletPicker handleWalletChange={handleWalletPicked} />
                </Modal>
            ) : ''}
        </Container>
    );
}

export default ExchangeItem;