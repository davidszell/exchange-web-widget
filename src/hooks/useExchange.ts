import { useEffect, useState } from 'react';
import { doExchange, fetchExchangeRates } from '../actions';
import { useAppDispatch, useAppSelector } from './reduxHooks';
import { ExchangeRatesType, WalletType } from '../types';
import { formatFloatRound } from '../helpers';

export type ExchangeActionType = 'buy' | 'sell';

const useExchange = () => {
  const wallets: WalletType[] = useAppSelector((state) => state.wallets);
  const exchangeRates: ExchangeRatesType = useAppSelector((state) => state.exchangeRates);

  const [sourceWallet, setSourceWallet] = useState<WalletType>(wallets[0]);
  const [destinationWallet, setDestinationWallet] = useState<WalletType>(wallets[1]);
  const [sourceAmount, setSourceAmount] = useState<number | null>(null);
  const [destinationAmount, setDestinationAmount] = useState<number | null>(null);
  const [exchangeAction, setExchangeAction] = useState<ExchangeActionType>('sell');
  const [isExchangeDisabled, setExchangeDisabled] = useState<boolean>(true);
  const [exchangeRate, setExchangeRate] = useState<number | null>(null);

  const dispatch = useAppDispatch();

  useEffect(() => {
    dispatch(fetchExchangeRates());

    const interval = setInterval(() => {
      dispatch(fetchExchangeRates());
    }, 10000);

    return () => clearInterval(interval);
  }, [dispatch]);

  useEffect(() => {
    setSourceWallet((prevWallet) => wallets.find((wallet) => wallet.name === prevWallet.name) || wallets[0]);
    setDestinationWallet((prevWallet) => wallets.find((wallet) => wallet.name === prevWallet.name) || wallets[1]);
  }, [wallets]);

  useEffect(() => {
    setExchangeDisabled(
      sourceAmount === null
      || destinationAmount === null
      || (exchangeAction === 'sell' && sourceAmount > sourceWallet.balance)
      || (exchangeAction === 'buy' && destinationAmount > destinationWallet.balance),
    );
  }, [exchangeAction, sourceAmount, sourceWallet.balance, destinationAmount, destinationWallet.balance]);

  useEffect(() => {
    if (Object.keys(exchangeRates.rates).length === 0) {
      setExchangeRate(null);
      return;
    }

    const sourceRate = exchangeRates.rates[sourceWallet.name];
    const destinationRate = exchangeRates.rates[destinationWallet.name];

    if (sourceRate == null || destinationRate == null) {
      setExchangeRate(null);
      return;
    }

    if (sourceWallet.name === exchangeRates.base) {
      setExchangeRate(exchangeRates.rates[destinationWallet.name]);
    } else {
      setExchangeRate(destinationRate / sourceRate);
    }
  }, [sourceWallet, destinationWallet, exchangeRates]);

  const clearAmounts = () => {
    setSourceAmount(null);
    setDestinationAmount(null);
  };

  const handleSourceWalletChange = (walletName: string) => {
    if (sourceWallet.name === walletName) {
      return;
    }

    const newWallet = wallets.find((item) => item.name === walletName);

    if (!newWallet) {
      return;
    }

    if (destinationWallet.name === walletName) {
      setDestinationWallet(sourceWallet);
    }

    setSourceWallet(newWallet);
    clearAmounts();
  };

  const handleDestinationWalletChange = (walletName: string) => {
    if (destinationWallet.name === walletName) {
      return;
    }

    const newWallet = wallets.find((item) => item.name === walletName);

    if (!newWallet) {
      return;
    }

    if (sourceWallet.name === walletName) {
      setSourceWallet(destinationWallet);
    }

    setDestinationWallet(newWallet);
    clearAmounts();
  };

  const handleSourceAmountChange = (newAmount: number | null) => {
    if (exchangeRate !== null && newAmount !== null && newAmount > 0) {
      setSourceAmount(newAmount);
      setDestinationAmount(formatFloatRound(newAmount * exchangeRate));
    } else {
      clearAmounts();
    }
  };

  const handleDestinationAmountChange = (newAmount: number | null) => {
    if (exchangeRate !== null && newAmount !== null && newAmount > 0) {
      setDestinationAmount(newAmount);
      setSourceAmount(formatFloatRound(newAmount / exchangeRate));
    } else {
      clearAmounts();
    }
  };

  const toggleExchangeAction = () => {
    setExchangeAction((prevExchangeAction) => (prevExchangeAction === 'sell' ? 'buy' : 'sell'));
  };

  const handleExchange = () => {
    if (!isExchangeDisabled) {
      dispatch(doExchange(sourceWallet.name, destinationWallet.name, exchangeAction, sourceAmount, destinationAmount));
      clearAmounts();
    }
  };

  return {
    exchangeAction,
    exchangeRate,
    isExchangeDisabled,
    sourceWallet,
    destinationWallet,
    sourceAmount,
    destinationAmount,
    handleExchange,
    handleSourceWalletChange,
    handleDestinationWalletChange,
    handleSourceAmountChange,
    handleDestinationAmountChange,
    toggleExchangeAction,
  };
};

export default useExchange;
