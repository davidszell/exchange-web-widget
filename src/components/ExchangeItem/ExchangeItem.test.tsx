import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import '@testing-library/jest-dom';
import { fireEvent, render, screen, within } from '@testing-library/react';
import ExchangeItem from './ExchangeItem';
import { WalletType } from '../../types';
import * as reduxHooks from '../../hooks/reduxHooks';

describe('ExchangeItem', () => {

  const useSelectorMock = jest.spyOn(reduxHooks, 'useAppSelector')
  const useDispatchMock = jest.spyOn(reduxHooks, 'useAppDispatch')

  beforeEach(() => {
    useSelectorMock.mockImplementation(selector => selector(mockStore));
    useDispatchMock.mockReturnValue(jest.fn())
  })
  afterEach(() => {
    useSelectorMock.mockClear()
    useDispatchMock.mockClear()
  })

  const mockStore = {
    wallets: [
      {
        name: "EUR",
        balance: 10,
        longName: "Euros",
        symbol: "€"
      },
      {
        name: "GBP",
        balance: 20,
        longName: "British Pounds",
        symbol: "£"
      },
      {
        name: "USD",
        balance: 30,
        longName: "US Dollars",
        symbol: "$"
      }
    ],
    exchangeRates: {
      base: "EUR",
      rates: {
        "EUR": 1,
        "GBP": 0.85,
        "USD": 1.16
      }
    }
  };

  it('matches snapshot', () => {
    const wallet: WalletType = {
      name: "EUR",
      balance: 10,
      longName: "Euros",
      symbol: "€"
    }
    const tree = renderer
      .create(<ExchangeItem deduct wallet={wallet} handleWalletChange={() => { }} amount={0} handleAmountChange={() => { }} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  })

  it('displays amount', () => {
    const wallet: WalletType = {
      name: "EUR",
      balance: 10,
      longName: "Euros",
      symbol: "€"
    }
    render(<ExchangeItem wallet={wallet} handleWalletChange={() => { }} amount={10} handleAmountChange={() => { }} />);
    const amountField = screen.getByTestId("amount");
    expect(amountField).toHaveDisplayValue('10');
  })

  it('calls handleAmountChange when amount is changed', () => {
    const wallet: WalletType = {
      name: "EUR",
      balance: 10,
      longName: "Euros",
      symbol: "€"
    }
    const handleAmountChangeMock = jest.fn();
    render(<ExchangeItem wallet={wallet} handleWalletChange={() => { }} amount={10} handleAmountChange={handleAmountChangeMock} />);
    fireEvent.change(screen.getByTestId('amount'), {target: {value: '20'}});
    expect(handleAmountChangeMock).toBeCalledWith(20);
  })

  it('displays modal when clicking on account picker', () => {
    const wallet: WalletType = {
      name: "EUR",
      balance: 10,
      longName: "Euros",
      symbol: "€"
    }
    render(<ExchangeItem wallet={wallet} handleWalletChange={() => { }} amount={0} handleAmountChange={() => { }} />);
    fireEvent.click(screen.getByText('EUR'));
    const modal = screen.getByTestId('modal');
    expect(modal).toBeDefined();
  })

  it('calls handleWalletChange when wallet is selected', () => {
    const wallet: WalletType = {
      name: "EUR",
      balance: 10,
      longName: "Euros",
      symbol: "€"
    }
    const handleWalletChangeMock = jest.fn();
    render(<ExchangeItem wallet={wallet} handleWalletChange={handleWalletChangeMock} amount={0} handleAmountChange={() => { }} />);
    fireEvent.click(screen.getByText('EUR'));
    const modal = screen.getByTestId('modal');
    fireEvent.click(within(modal).getByText('USD'));
    expect(handleWalletChangeMock).toBeCalledWith('USD');
  })

  it('can close modals', () => {
    const wallet: WalletType = {
      name: "EUR",
      balance: 10,
      longName: "Euros",
      symbol: "€"
    }
    render(<ExchangeItem wallet={wallet} handleWalletChange={() => { }} amount={0} handleAmountChange={() => { }} />);
    fireEvent.click(screen.getByText('EUR'));
    const modal = screen.getByTestId('modal');
    fireEvent.click(within(modal).getByText(/x/i));
    expect(screen.queryByTestId('modal')).toBeNull();
  })

  it('shows error when deducting and amount is larger than available balance', () => {
    const wallet: WalletType = {
      name: "EUR",
      balance: 10,
      longName: "Euros",
      symbol: "€"
    }
    render(<ExchangeItem deduct wallet={wallet} handleWalletChange={() => { }} amount={11} handleAmountChange={() => { }} />);
    expect(screen.getByText('Exceeds balance')).not.toBeNull();
  })

  it('does not show error when not deducting and amount is larger than available balance', () => {
    const wallet: WalletType = {
      name: "EUR",
      balance: 10,
      longName: "Euros",
      symbol: "€"
    }
    render(<ExchangeItem wallet={wallet} handleWalletChange={() => { }} amount={11} handleAmountChange={() => { }} />);
    expect(screen.queryByText('Exceeds balance')).toBeNull();
  })
})