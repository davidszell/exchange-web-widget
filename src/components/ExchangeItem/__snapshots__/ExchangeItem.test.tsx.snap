// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExchangeItem matches snapshot 1`] = `
.c4 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  border: none;
  text-align: right;
  font-size: 1.3em;
  padding: 10px;
  border-radius: 5px;
  -moz-appearance: textfield;
}

.c4::-webkit-outer-spin-button,
.c4::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

.c2 {
  border: 1px solid #157ae2;
  color: #157ae2;
  background-color: white;
  padding: 10px;
  border-radius: 5px;
  cursor: pointer;
  font-weight: 500;
}

.c2:hover {
  color: white;
  background-color: #157ae2;
}

.c2:disabled,
.c2[disabled],
.c2:disabled:hover,
.c2[disabled]:hover {
  cursor: unset;
  color: #777;
  border-color: #777;
  background-color: white;
}

.c0 {
  padding: 10px;
  border: 1px solid #ccc;
  border-radius: 5px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 10px;
}

.c0.error {
  border-color: #e21534;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  gap: 10px;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c3::after {
  content: '⮁';
  margin-left: 10px;
}

.c6 {
  font-size: 0.8em;
  margin: 0;
  color: #777;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <button
      className="c2 c3"
      onClick={[Function]}
    >
      EUR
    </button>
    <input
      className="c4"
      data-testid="amount"
      onChange={[Function]}
      placeholder="0"
      step="0.01"
      type="number"
      value=""
    />
  </div>
  <div
    className="c5"
  >
    <p
      className="c6"
    >
      Balance:
      €
      10
    </p>
  </div>
</div>
`;
